### PR TITLE
Callback Testing API

### PR DESCRIPTION
**Pull request by danielmarbach**
Candidate for https://github.com/Particular/NServiceBus.Callbacks/issues/48

@Particular/nservicebus-maintainers what do you think?

We could release this alongside with the callbacks as NServiceBus.Callbacks.Testing nuget

### CODE REVIEW

**Comment by danielmarbach** [comment]
@timbussmann this proofs also that our testable context approach rocks *shameless plug*

**Comment by timbussmann** [comment]
> @timbussmann this proofs also that our testable context approach rocks shameless plug

🎸 💃 

I like the approach. I was thinking at something similar, but instead of defining the when conditions upfront to give access to the TaskCompletionSource. Usage then could have looked something like:

```
var request = session.Request<>(...);
session.SetCallbackResult(...);
await request;
```

but this introduces quite a complexity about the order of execution and everything, which solved better with your suggestion 👍

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers this is ready for review. Will add a doco PR soon

**Comment by ramonsmits** [comment]
Nice job @danielmarbach , this does not work within handlers/sagas right? Unless you probably inject `IMessageSession` but you should not be doing that, just wondering. It's is very close to mocking if the types are visible as typed in https://github.com/Particular/NServiceBus.Callbacks/issues/48#issuecomment-259083357

```c#
var messageSession = new TestableCallbackAwareSession();
messageSession.When<Request, string>(r =>true, 42);
var result = await messageSession.Request<string>(new Request());
Assert.AreEqual(42, result);
```

Wondering what would be needed if there are multiple extensions besides callbacks and all would be used in a single test. It's probably YAGNI.

**Comment by danielmarbach** [comment]
@ramonsmits we do not support callbacks in handlers and sagas ;)

I thought about a more compositional approach where I write extension methods on top of TestableMessageSession. It would be doable. The trick would be the have a private nested session which decorates the extended session but then you'd need to make sure you use that and pass it into the code that you want to test. I think it would be cumbersome to use and defeat the purpose.

File: packaging/nuget/nservicebus.callbacks.testing.nuspec [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    <description>The most popular open-source service bus for .net</description>
+    <releaseNotes></releaseNotes>
+    <copyright>$copyright$</copyright>
+    <tags>$tags$</tags>
+    <frameworkAssemblies>
+      <frameworkAssembly assemblyName="System" targetFramework="net452" />
+      <frameworkAssembly assemblyName="System.Core" targetFramework="net452" />
+    </frameworkAssemblies>
+    <dependencies>
+      <dependency id="NServiceBus.Callbacks" version="$version$" />
```
Comment (timbussmann) [inline packaging/nuget/nservicebus.callbacks.testing.nuspec]: don't we need a dependency to `NServiceBus.Testing` ?

Comment (danielmarbach) [inline packaging/nuget/nservicebus.callbacks.testing.nuspec]: Correct. I was sure I added that but the history proves otherwise 🗡

**Comment by danielmarbach** [comment]
Addressed @timbussmann comments

**Comment by hlubovac** [comment]
A suggestion: second argument of `When` would be better of being a `Func<TResponse>` or `Func<TRequest, TResponse>`, I think. I don't see a way to test cases involving errors raised within (those that don't involve functionality of NServiceBus; e.g. `AggregateException` holding a `TaskCanceledException`).

**Pull request merged**